File: priority_queue/priority_queue_via_max_heap_array.cpp
/**
 * @file priority_queue_via_max_heap_array.cpp
 * @brief C++ Program to implement Priority Queue using Max Heap
 */
//

#include <iostream>
#include <vector>
#include <stdexcept>

template <typename PriorityQueueElementType>
class PriorityQueueViaMaxHeapArray {
private:
    std::vector<PriorityQueueElementType> maxHeapArray;
    unsigned int size = 0;

    void heapifyUp(unsigned int index);             // Helper method to maintain the heap property after insertion
    void heapifyDown(unsigned int index);           // Helper method to maintain the heap property after removal
    
public:
    void insert(PriorityQueueElementType data);     // Method to insert an element into the priority queue
    PriorityQueueElementType remove();              // Method to remove the element with the highest priority (highest value) from the priority queue
    PriorityQueueElementType getMax();              // Method to get the element with the highest priority (highest value) from the priority queue
    PriorityQueueElementType getMin();              // Method to get the element with the lowest priority (lowest value) from the priority queue
    unsigned int getSize();                         // Method to get the this->size of the priority queue
    void display();                                 // Method to display the elements of the priority queue via level order traversal
};

// Method to insert an element into the priority queue
template <typename PriorityQueueElementType>
void PriorityQueueViaMaxHeapArray<PriorityQueueElementType>::insert(PriorityQueueElementType data) {
    maxHeapArray.push_back(data);
    this->size++;
    heapifyUp(this->size - 1);
}

// Helper method to maintain the heap property after insertion
template <typename PriorityQueueElementType>
void PriorityQueueViaMaxHeapArray<PriorityQueueElementType>::heapifyUp(unsigned int index) {
    while (index > 0) {
        // Note that the max heap is implemented as an array, so the parent index can be calculated as (index - 1) / 2
        unsigned int parentIndex = (index - 1) / 2;
        // If the parent node is less than the current node, swap them to maintain the max heap property
        if (maxHeapArray[parentIndex] < maxHeapArray[index]) {
            std::swap(maxHeapArray[parentIndex], maxHeapArray[index]);
            index = parentIndex;
        } else {
            break;
        }
    }
}

// Method to remove the element with the highest priority (highest value) from the priority queue
template <typename PriorityQueueElementType>
PriorityQueueElementType PriorityQueueViaMaxHeapArray<PriorityQueueElementType>::remove() {
    if (this->size == 0) {
        throw std::out_of_range("Priority Queue is empty");
    }

    // The element with the highest priority is always the root of the max heap
    // So, removing the root element will be removing the first element of the array
    PriorityQueueElementType removedElement = maxHeapArray[0];
    maxHeapArray[0] = maxHeapArray[this->size - 1];                 // Replace the root with the last element of the array
    maxHeapArray.pop_back();                                        // Remove the last element of the array (which was the root)
    this->size--;

    // Maintain the heap property after removal
    // Note that the root element is replaced with the last element of the array (which was the last element of the max heap)
    heapifyDown(0);
    return removedElement;
}

// Helper method to maintain the heap property after removal
template <typename PriorityQueueElementType>
void PriorityQueueViaMaxHeapArray<PriorityQueueElementType>::heapifyDown(unsigned int index) {
    while (index < this->size) {
        // Note that the max heap is implemented as an array, so the left child index can be calculated as 2 * index + 1
        unsigned int leftChildIndex = 2 * index + 1;
        unsigned int rightChildIndex = 2 * index + 2;
        unsigned int maxIndex = index;                      // Assume the current node is the largest

        if (leftChildIndex < this->size && maxHeapArray[leftChildIndex] > maxHeapArray[maxIndex]) {
            // If the left child is greater than the current node, update the max index
            maxIndex = leftChildIndex;
        }
        if (rightChildIndex < this->size && maxHeapArray[rightChildIndex] > maxHeapArray[maxIndex]) {
            // If the right child is greater than the current node, update the max index
            maxIndex = rightChildIndex;
        }
        if (maxIndex == index) {
            // If the current node is the largest, we don't need to do anything because the max heap property is maintained
            break;
        } else {
            // Otherwise, swap the current node with the largest child
            std::swap(maxHeapArray[index], maxHeapArray[maxIndex]);
            index = maxIndex;
        }
    }
}

// Method to get the element with the highest priority (highest value) from the priority queue
template <typename PriorityQueueElementType>
PriorityQueueElementType PriorityQueueViaMaxHeapArray<PriorityQueueElementType>::getMax() {
    if (this->size == 0) {
        throw std::out_of_range("Priority Queue is empty");
    }

    // The element with the highest priority is always the root of the max heap
    return maxHeapArray[0];
}

// Method to get the element with the lowest priority (lowest value) from the priority queue
template <typename PriorityQueueElementType>
PriorityQueueElementType PriorityQueueViaMaxHeapArray<PriorityQueueElementType>::getMin() {
    if (this->size == 0) {
        throw std::out_of_range("Priority Queue is empty");
    }
    
    // We just need to iterate through the max heap array to find the minimum element
    PriorityQueueElementType minElement = maxHeapArray[0];
    for (const auto& element : maxHeapArray) {
        if (element < minElement) {
            minElement = element;
        }
    }
    return minElement;
}

// Method to get the this->size of the priority queue
template <typename PriorityQueueElementType>
unsigned int PriorityQueueViaMaxHeapArray<PriorityQueueElementType>::getSize() {
    return this->size;
}

// Method to display the elements of the priority queue via level order traversal
template <typename PriorityQueueElementType>
void PriorityQueueViaMaxHeapArray<PriorityQueueElementType>::display() {
    if (this->size == 0) {
        throw std::out_of_range("Priority Queue is empty");
    }
    for (unsigned int index = 0; index < this->size; index++) {
        std::cout << maxHeapArray[index] << " ";
    }
    std::cout << std::endl;
}

// Main function
int main() {
    PriorityQueueViaMaxHeapArray<int> priorityQueue;

    int elementsToInsert[] = { 10, 20, 15, 40, 50, 100 };
    for (const auto& element : elementsToInsert) {
        priorityQueue.insert(element);
        priorityQueue.display();
    }

    std::cout << "Element with the highest priority: " << priorityQueue.getMax() << std::endl;
    std::cout << "Element with the lowest priority: " << priorityQueue.getMin() << std::endl;

    std::cout << "Removed element with the highest priority: " << priorityQueue.remove() << std::endl;
    priorityQueue.display();

    return 0;
}

// 10 
// 20 10 
// 20 10 15 
// 40 20 15 10 
// 50 40 15 10 20
// 100 40 50 10 20 15
// Element with the highest priority: 100
// Element with the lowest priority: 10
// Removed element with the highest priority: 100
// 50 40 15 10 20

File: priority_queue/priority_queue_via_max_heap_array.py
# file: priority_queue/priority_queue_via_max_heap_array.py
# brief: Priority Queue implementation using Max Heap
#        Max heap is a binary tree where the parent node is greater than its children
#        Max heap can be implemented using an array where the parent node is at index i

from typing import List, Union

class PriorityQueueViaMaxHeapArray:
    def __init__(self) -> None:
        self.heap:List[Union[int, float]] = []

    # Insert a value in the heap
    def insert(self, value:Union[int, float]) -> None:
        self.heap.append(value)
        self._heapify_up()

    # _heapify_up is a helper function to maintain the heap property while inserting a value
    def _heapify_up(self) -> None:
        index:int = len(self.heap) - 1
        while index > 0:
            parent_index:int = (index - 1) // 2
            if self.heap[index] > self.heap[parent_index]:
                # Swap the parent and child node if the child node is greater than the parent node (to keep the max heap property)
                self.heap[index], self.heap[parent_index] = self.heap[parent_index], self.heap[index]
                index = parent_index
            else:
                break

    # Remove the maximum value from the heap
    def remove(self) -> Union[int, float]:
        if len(self.heap) == 0:
            raise ValueError("Heap is empty")
        max_value = self.heap[0]
        self.heap[0] = self.heap[-1]
        self.heap.pop()
        self._heapify_down()
        return max_value
    
    # _heapify_down is a helper function to maintain the heap property while removing the maximum value
    def _heapify_down(self) -> None:
        index:int = 0
        while index < len(self.heap):
            # Find the left and right child index of the current node.
            # Note that this implementation uses an array to represent the heap
            left_child_index:int = 2 * index + 1
            right_child_index:int = 2 * index + 2
            if left_child_index >= len(self.heap):
                # If the left child index is greater than the heap size, then the current node is a leaf node
                break
            if right_child_index >= len(self.heap):
                # If the right child index is greater than the heap size, then the current node has only one child
                right_child_index = left_child_index
            
            max_child_index:int = left_child_index if self.heap[left_child_index] > self.heap[right_child_index] else right_child_index
            if self.heap[index] < self.heap[max_child_index]:
                # Swap the parent and child node if the child node is greater than the parent node (to keep the max heap property)
                self.heap[index], self.heap[max_child_index] = self.heap[max_child_index], self.heap[index]
                # Change the index to the child node index
                index = max_child_index
            else:
                break

    # Get the maximum value from the heap
    def get_max(self) -> Union[int, float]:
        if len(self.heap) == 0:
            raise ValueError("Heap is empty")
        return self.heap[0]
    
    # Get the minimum value from the heap
    def get_min(self) -> Union[int, float]:
        if len(self.heap) == 0:
            raise ValueError("Heap is empty")
        # We just need to iterate through the heap(array) to find the minimum value
        min_value = self.heap[0]
        for value in self.heap:
            if value < min_value:
                min_value = value
        return min_value
    
    # Get the size of the heap
    def size(self) -> int:
        return len(self.heap)
    
    # Print the heap via inorder traversal
    def inorder_traversal(self) -> None:
        self._inorder_traversal(0)

    # Helper function to print the heap via inorder traversal
    def _inorder_traversal(self, index:int) -> None:
        if index >= len(self.heap):
            return
        self._inorder_traversal(2 * index + 1)
        print(self.heap[index], end=" ")
        self._inorder_traversal(2 * index + 2)
        return
    
    # Print the heap via level order traversal
    def level_order_traversal(self) -> None:
        for value in self.heap:
            print(value, end=" ")
        return
    
# Usage
if __name__ == "__main__":
    pq = PriorityQueueViaMaxHeapArray()

    pq.insert(10)
    pq.insert(20)
    pq.insert(15)
    pq.insert(40)
    pq.insert(50)
    pq.insert(100)
    pq.insert(25)
    pq.insert(45)
    pq.insert(30)
    pq.insert(35)
    pq.insert(5)
    pq.insert(60)
    pq.insert(55)
    pq.insert(70)
    pq.insert(80)
    pq.insert(90)
    pq.insert(75)
    pq.insert(65)
    pq.insert(85)
    pq.insert(95)
    
    pq.level_order_traversal()
    print()
    print("Max value:", pq.get_max())
    print("Min value:", pq.get_min())
    print("Size:", pq.size())
    print("Removing max value:", pq.remove())
    print("Max value:", pq.get_max())
    print("Min value:", pq.get_min())
    print("Size:", pq.size())
    pq.level_order_traversal()

# 100 95 80 85 90 55 70 45 75 35 5 15 50 25 60 10 40 30 65 20 
# Max value: 100
# Min value: 5
# Size: 20
# Removing max value: 100
# Max value: 95
# Min value: 5
# Size: 19
# 95 90 80 85 35 55 70 45 75 20 5 15 50 25 60 10 40 30 65